src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnusedCatchParameterShouldBeUnnamedCheck.java

```diff
@@ -0,0 +1,270 @@
+ *     </li>
+ *     <li>
+ *         Follows Java conventions for denoting unused parameters with an underscore ({@code _}).
+ *     </li>
+ * </ul>
+ * <p>
+ * See the <a href="https://docs.oracle.com/en/java/javase/21/docs/specs/unnamed-jls.html">
+ * Java Language Specification</a> for more information about unnamed variables.
+ * </p>
+ * <p>
+ * <b>Attention</b>: This check should be activated only on source code
+ * that is compiled by jdk21 or higher;
+ * unnamed catch parameters came out as the first preview in Java 21.
+ * </p>
+ * <p>
+ * Parent is {@code com.puppycrawl.tools.checkstyle.TreeWalker}
+ * </p>
+ * <p>
+ * Violation Message Keys:
+ * </p>
+ * <ul>
+ * <li>
+ * {@code unused.catch.parameter}
+ * </li>
+ * </ul>
+ *
+ * @since 10.18.0
+ *
+ */
+
+@FileStatefulCheck
+public class UnusedCatchParameterShouldBeUnnamedCheck extends AbstractCheck {
+
+    /**
+     * A key is pointing to the warning message text in "messages.properties"
+     * file.
+     */
+    public static final String MSG_UNUSED_CATCH_PARAMETER = "unused.catch.parameter";
+
+    /**
+     * Invalid parents of the catch parameter identifier.
+     */
+    private static final int[] INVALID_CATCH_PARAM_IDENT_PARENTS = {
+        TokenTypes.DOT,
+        TokenTypes.LITERAL_NEW,
+        TokenTypes.METHOD_CALL,
+        TokenTypes.TYPE,
+    };
+
+    /**
+     * Keeps track of the catch parameters in a block.
+     */
+    private final Deque<CatchParameterDetails> catchParameters = new ArrayDeque<>();
+
+    @Override
+    public int[] getDefaultTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getAcceptableTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getRequiredTokens() {
+        return new int[] {
+            TokenTypes.LITERAL_CATCH,
+            TokenTypes.IDENT,
+        };
+    }
+
+    @Override
+    public void visitToken(DetailAST ast) {
+        if (ast.getType() == TokenTypes.LITERAL_CATCH) {
+            final CatchParameterDetails catchParameter = new CatchParameterDetails(ast);
+            catchParameters.push(catchParameter);
+        }
+        else if (isCatchParameterIdentifierCandidate(ast) && !isLeftHandOfAssignment(ast)) {
+            // we do not count reassignment as usage
+            catchParameters.stream()
+                    .filter(parameter -> parameter.getName().equals(ast.getText()))
+                    .findFirst()
+                    .ifPresent(CatchParameterDetails::registerAsUsed);
+        }
+    }
+
+    @Override
+    public void leaveToken(DetailAST ast) {
+        final CatchParameterDetails catchParameter = catchParameters.peek();
+        if (catchParameter != null && ast.equals(catchParameter.getEnclosingCatchClause())) {
+            if (!"_".equals(catchParameter.getName()) && !catchParameter.isUsed()) {
+                log(catchParameter.getParameterDefinition(),
```

Comment:
I don't know what is best practice here (logging on the name or the parameter def token). but I tried to keep consistency with `UnusedLocalVariableCheck`.[ It logs on the type of the variable](https://github.com/checkstyle/checkstyle/blob/15b5b73406f9ec8311625bb26baf8530ef8eca4d/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnusedLocalVariableCheck.java#L427) not its name (ident)

Do you think it will be more precise it we log on the ident token ?

Reply:
I am not sure this has really come up.

I would expect we send the AST which is being violated as bad. Maybe some special filter will look at this AST and do more processing. Parameter def and it's ident probably have the same line/column, so most of our processes won't care.

> UnusedLocalVariableCheck.[ It logs on the type of the variable](https://github.com/checkstyle/checkstyle/blob/15b5b73406f9ec8311625bb26baf8530ef8eca4d/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnusedLocalVariableCheck.java#L427) not its name (ident)

That is odd and doesn't make any sense. This has a different column number. At least your current makes a bit more sense.

@nrmancuso What do you think?

Reply:
If we are calling the check `...Catchparameter...`, we should place violations on the catch parameter, not just the identifier. Grammar for reference: https://docs.oracle.com/javase/specs/jls/se21/html/jls-14.html#jls-CatchFormalParameter

```
TryStatement:
  try Block Catches
  try Block [Catches] Finally

TryWithResourcesStatement
  Catches:
  CatchClause {CatchClause}

CatchClause:
  catch ( CatchFormalParameter ) Block

CatchFormalParameter:
  {VariableModifier} CatchType VariableDeclaratorId

CatchType:
  UnannClassType {| ClassType}

Finally:
  finally Block
```

Reply:
@mahfouz72 please add some modifiers in front of some of the types (final) to make sure that we are indeed placing violations on the right thing.

Reply:
Done.  

```
        try {
            int x = 1/ 0;
        }
        catch (final Exception e) {  // violation

        }
        -LITERAL_CATCH -> catch [17:8]
|       |       |       |--LPAREN -> ( [17:14]
|       |       |       |--PARAMETER_DEF -> PARAMETER_DEF [17:15]  <--- we log this
|       |       |       |   |--MODIFIERS -> MODIFIERS [17:15]
|       |       |       |   |   `--FINAL -> final [17:15]
|       |       |       |   |--TYPE -> TYPE [17:21]
|       |       |       |   |   `--IDENT -> Exception [17:21]
|       |       |       |   `--IDENT -> e [17:31]
```